src/reader/async_tokio.rs

```diff
@@ -81,19 +81,52 @@ impl<R: AsyncBufRead + Unpin> Reader<R> {
     /// ```
     ///
     /// [`read_event_into()`]: Reader::read_event_into
-    pub fn read_event_into_async<'reader, 'b: 'reader>(
+    pub async fn read_event_into_async<'reader, 'b: 'reader>(
         &'reader mut self,
         buf: &'b mut Vec<u8>,
-    ) -> Pin<Box<dyn Future<Output = Result<Event<'b>>> + 'reader>> {
-        Box::pin(async move {
-            read_event_impl!(
-                self, buf,
-                TokioAdapter(&mut self.reader),
-                read_until_open_async,
-                read_until_close_async,
-                await
-            )
-        })
+    ) -> Result<Event<'b>> {
+        let event = self.read_event_into_async_inner(buf).await;
+        match event {
+            Err(_) | Ok(Event::Eof) => self.parser.state = ParseState::Exit,
+            _ => {}
+        }
+        event
+    }
+
+    async fn read_event_into_async_inner<'reader, 'b: 'reader>(
+        &'reader mut self,
+        mut buf: &'b mut Vec<u8>,
+    ) -> Result<Event<'b>> {
+        loop {
+            match self.parser.state {
+                ParseState::Init => {
+                    #[cfg(feature = "encoding")]
+                    if let Some(encoding) =
+                        (TokioAdapter(&mut self.reader)).detect_encoding().await?
+                    {
+                        if self.parser.encoding.can_be_refined() {
+                            self.parser.encoding =
+                                crate::reader::EncodingRef::BomDetected(encoding);
+                        }
+                    }
+                    #[cfg(not(feature = "encoding"))]
+                    (TokioAdapter(&mut self.reader)).remove_utf8_bom().await?;
+                    match self.read_until_open_async(buf).await? {
+                        Ok(ev) => break Ok(ev),
+                        Err(b) => buf = b,
```

Comment:
I should probably add that this trick is to circumvent [this limitation](https://github.com/rust-lang/rfcs/blob/master/text/2094-nll.md#problem-case-3-conditional-control-flow-across-functions) in pre-non-lexical borrowck